targets/TARGET_STM/i2c_api.c

```diff
@@ -61,24 +59,40 @@
 #define I2C_NUM (5)
 static I2C_HandleTypeDef* i2c_handles[I2C_NUM];
 
-/* Timeout values are based on core clock and I2C clock.
-   The BYTE_TIMEOUT is computed as twice the number of cycles it would
+/* Timeout value based on Core and I2C clocks */
+#define BYTE_TIMEOUT (SystemCoreClock / obj_s->hz)
+
+/* The BYTE_TIMEOUT_20 is computed as twice the number of cycles it would
    take to send 10 bits over I2C. Most Flags should take less than that.
    This is for immediate FLAG or ACK check.
 */
-#define BYTE_TIMEOUT ((SystemCoreClock / obj_s->hz) * 2 * 10)
-/* Timeout values based on I2C clock.
-   The BYTE_TIMEOUT_US is computed as 3x the time in us it would
+#define BYTE_TIMEOUT_20 (BYTE_TIMEOUT * 2 * 10)
+
+/* The BYTE_TIMEOUT_30 is computed as 3x the time in us it would
    take to send 10 bits over I2C. Most Flags should take less than that.
    This is for complete transfers check.
 */
-#define BYTE_TIMEOUT_US   ((SystemCoreClock / obj_s->hz) * 3 * 10)
+#define BYTE_TIMEOUT_30 (BYTE_TIMEOUT * 3 * 10)
+
 /* Timeout values for flags and events waiting loops. These timeouts are
    not based on accurate values, they just guarantee that the application will
    not remain stuck if the I2C communication is corrupted. 
 */
 #define FLAG_TIMEOUT ((int)0x1000)
 
+/* Instead of using wait_us() which can cause issues in some cases.
+   This function assumes the tick is 1 us.
+*/
+static void wait_loop_us(uint32_t timeout)
+{
+    uint32_t t1, t2, elapsed = 0;
+    t1 = us_ticker_read();
+    do {
+        t2 = us_ticker_read();
+        elapsed = (t2 > t1) ? (t2 - t1) : ((uint64_t)t2 + 0xFFFFFFFF - t1 + 1);
```

Comment:
All the above isn't necessary - just `elapsed = t2 - t1;` will suffice even on wrap because of the way unsigned arithmetic works.

Reply:
OK. Thanks for the review. I will update it.

Reply:
are you sure ? we don't know if the ticker is 32 bits or 16 bits wide ...

Reply:
I wasn't aware the ticker could be less than 32 bits wide, but seems it can.

But if it can be, that wrap logic wouldn't be correct. It's adding 2^32 when it sees it wrap. It should be adding 2^(ticker width), right?

I've seen similar wrap logic in other places in the tree, with both 2^16 and 2^32 increments, apparently with hard knowledge of their timer width. For this platform, it seems to be 32-bit.

Reply:
You're right ! the bit width of timer needs to be used. In STM32 targets we have both 32-bits and 16-bits timers and the I2C driver is common for all ... I definitely think that this function should be implemented carefully and would better belong to common code in mbed hal common layer for instance.

Reply:
If it can be 16-bit for you, I think you should be implementing `us_ticker_get_info()` to reveal that to the world, which I don't see in any STM32 target directory. That is relied on to get the wrap right in some existing code.

Double-checking, looks like you implement your own 16->32-bit expansion in us_ticker_16b.c. So your `us_ticker_read` is always 32-bit.

Reply:
The implementation of us_ticker_get_info() is done in ticker feature branch. This should land on master very soon. So this needs to be taken into account from now on.